src/common/src/array/value_reader.rs

```diff
@@ -57,31 +54,27 @@ impl_numeric_value_reader!(OrderedF32, F32ValueReader, read_f32);
 impl_numeric_value_reader!(OrderedF64, F64ValueReader, read_f64);
 
 pub trait VarSizedValueReader<AB: ArrayBuilder> {
-    fn read(buf: &[u8]) -> Result<<<AB as ArrayBuilder>::ArrayType as Array>::RefItem<'_>>;
+    fn read(buf: &[u8]) -> ArrayResult<<<AB as ArrayBuilder>::ArrayType as Array>::RefItem<'_>>;
 }
 
 pub struct Utf8ValueReader {}
 
 impl VarSizedValueReader<Utf8ArrayBuilder> for Utf8ValueReader {
-    fn read(buf: &[u8]) -> Result<&str> {
-        from_utf8(buf).map_err(|e| {
-            RwError::from(InternalError(format!(
-                "failed to read utf8 string from bytes: {}",
-                e
-            )))
-        })
+    fn read(buf: &[u8]) -> ArrayResult<&str> {
+        match from_utf8(buf) {
+            Ok(s) => Ok(s),
+            Err(e) => bail!("failed to read utf8 string from bytes: {}", e),
+        }
     }
 }
 
 pub struct DecimalValueReader {}
 
 impl VarSizedValueReader<DecimalArrayBuilder> for DecimalValueReader {
-    fn read(buf: &[u8]) -> Result<Decimal> {
-        Decimal::from_str(Utf8ValueReader::read(buf)?).map_err(|e| {
-            RwError::from(InternalError(format!(
-                "failed to read decimal from string: {}",
-                e
-            )))
-        })
+    fn read(buf: &[u8]) -> ArrayResult<Decimal> {
+        match Decimal::from_str(Utf8ValueReader::read(buf)?) {
+            Ok(d) => Ok(d),
+            Err(e) => bail!("failed to read decimal from string: {}", e),
```

Comment:
It would be better to use anyhow's context function to attach extra info to an existing error.